rs/bitcoin/ckbtc/minter/src/fees/mod.rs

```diff
@@ -0,0 +1,122 @@
+use crate::state::CkBtcMinterState;
+use crate::tx::UnsignedTransaction;
+use crate::{Network, fake_sign};
+use ic_btc_interface::{MillisatoshiPerByte, Satoshi};
+use std::cmp::max;
+
+pub trait FeeEstimator {
+    fn estimate_median_fee(
+        &self,
+        fee_percentiles: &[MillisatoshiPerByte],
+    ) -> Option<MillisatoshiPerByte>;
+
+    /// Evaluate the fee necessary to cover the minter's cycles consumption.
+    fn evaluate_minter_fee(&self, num_inputs: u64, num_outputs: u64) -> Satoshi;
+
+    /// Evaluate transaction fee with the given fee rate (in milli base unit per vbyte/byte)
+    fn evaluate_transaction_fee(&self, tx: &UnsignedTransaction, fee_rate: u64) -> u64;
+
+    /// Compute a new minimum withdrawal amount based on the current fee rate
+    fn fee_based_minimum_withrawal_amount(&self, median_fee: MillisatoshiPerByte) -> Satoshi;
+}
+
+pub struct BitcoinFeeEstimator {
+    network: Network,
+    retrieve_btc_min_amount: u64,
+    check_fee: u64,
+}
+
+impl BitcoinFeeEstimator {
+    pub fn new(network: Network, retrieve_btc_min_amount: u64, check_fee: u64) -> Self {
+        Self {
+            network,
+            retrieve_btc_min_amount,
+            check_fee,
+        }
+    }
+
+    pub fn from_state(state: &CkBtcMinterState) -> Self {
+        Self::new(
+            state.btc_network,
+            state.retrieve_btc_min_amount,
+            state.check_fee,
+        )
+    }
+
+    /// An estimated fee per vbyte of 142 millistatoshis per vbyte was selected around 2025.06.21 01:09:50 UTC
+    /// for Bitcoin Mainnet, whereas the median fee around that time should have been 2_000.
+    /// Until we know the root cause, we ensure that the estimated fee has a meaningful minimum value.
+    const fn minimum_fee_per_vbyte(&self) -> MillisatoshiPerByte {
+        match &self.network {
+            Network::Mainnet => 1_500,
+            Network::Testnet => 1_000,
+            Network::Regtest => 0,
+        }
+    }
+}
+
+impl FeeEstimator for BitcoinFeeEstimator {
+    fn estimate_median_fee(
+        &self,
+        fee_percentiles: &[MillisatoshiPerByte],
+    ) -> Option<MillisatoshiPerByte> {
+        /// The default fee we use on regtest networks.
+        const DEFAULT_REGTEST_FEE: MillisatoshiPerByte = 5_000;
+
+        let median_fee = match &self.network {
+            Network::Mainnet | Network::Testnet => {
+                if fee_percentiles.len() < 100 {
+                    return None;
+                }
+                Some(fee_percentiles[50])
+            }
+            Network::Regtest => Some(DEFAULT_REGTEST_FEE),
```

Comment:
Is there a special casing for `Regtest` because there may not be sufficiently many transactions when working with a local setup?